File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.java
class Solution
{
    public int solution(int [][]board)
    {
        int N = board.length;
        int M = board[0].length;
        
        int answer = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (board[i][j] != 0) {
                    if (answer == 0) answer = 1;
                    if (i - 1 >= 0 && j - 1 >= 0) {
                        int tmp = Math.min(board[i-1][j-1], board[i-1][j]);
                        tmp = Math.min(tmp, board[i][j-1]);
                        board[i][j] += tmp;
                        answer = Math.max(answer, board[i][j]);
                    }
                }
                
            }
        }
        // for (int i = 0; i < N; i++) {
        //     for (int j = 0; j < M; j++) {
        //         System.out.print(board[i][j] + " ");
        //     }
        //     System.out.println();
        // }

        return answer * answer;
    }
}

File: 프로그래머스/2/12905. 가장 큰 정사각형 찾기/가장 큰 정사각형 찾기.py
def solution(board):
    answer = 0
    N = len(board)
    M = len(board[0])
    
    for i in range(N):
        for j in range(M):
            if board[i][j] == 1:
                if answer == 0:
                    answer = 1
                    
                if i-1 >= 0 and j-1 >= 0:
                    board[i][j] += min(board[i-1][j-1], board[i-1][j], board[i][j-1])
                    answer = max(answer, board[i][j])

    return answer * answer